vllm/model_executor/layers/quantization/marlin.py

```diff
@@ -0,0 +1,188 @@
+        output_size_per_partition: int,
+        input_size: int,
+        output_size: int,
+        params_dtype: torch.dtype,
+    ) -> Dict[str, Any]:
+        del output_size  # Unused.
+        if params_dtype != torch.float16:
+            raise ValueError(
+                f"The params dtype must be float16, but got {params_dtype}")
+        if input_size_per_partition % self.quant_config.group_size != 0:
+            raise ValueError(
+                "The input size is not aligned with the quantized "
+                "weight shape. This can be caused by too large "
+                "tensor parallel size.")
+        if output_size_per_partition % self.quant_config.pack_factor != 0:
+            raise ValueError(
+                "The output size is not aligned with the quantized "
+                "weight shape. This can be caused by too large "
+                "tensor parallel size.")
+        if input_size_per_partition % 128 != 0:
+            raise ValueError(
+                "The input_size_per_partition must be divisible by 128, "
+                f"but got {input_size_per_partition}")
+        
+        if output_size_per_partition % 256 != 0:
+            raise ValueError(
+                "The output_size_per_partition must be divisible by 256, "
+                f"but got {output_size_per_partition}")
+
+        # check that we have at least 4 tiles horizontally in the shard
+        num_tiles_per_perm = self._perm_len // (self.quant_config.tile_size ** 2)
+        if output_size_per_partition % num_tiles_per_perm != 0:
+            raise ValueError(
+                "Each permutation group must reside on the same gpu"
+            )
+
+        # Quantized 4Bit weights packed into Int32.
+        qweight = Parameter(
+            torch.empty(
+                input_size_per_partition // self.quant_config.tile_size, 
+                output_size_per_partition * self.quant_config.tile_size // self.quant_config.pack_factor, 
+                device="cuda",
+                dtype=torch.int32
+            ),
+            requires_grad=False,
+        )
+        set_weight_attrs(
+            qweight, {
+                "input_dim": 0,
+                "output_dim": 1,
+                "packed_dim": 1,
+                "pack_factor": self.quant_config.pack_factor,
+                "tile_size": TILE_SIZE,
+            })
+
+        # Scales in Float16.
+        scales = Parameter(
+            torch.empty(
+                input_size_per_partition // self.quant_config.group_size,
+                output_size_per_partition, 
+                device="cuda",
+                dtype=params_dtype,
+            ),
+            requires_grad=False,
+        )
+        set_weight_attrs(scales, {
+            "input_dim": None if input_size == input_size_per_partition else 0,
+            "output_dim": 1,
+        })
+
+        # Workspace for the marlin kernels.
+        self.workspace = torch.empty(MAX_SMS, dtype=torch.int)
```

Comment:
Since we have set `256` here for `workspace`. If we really don’t need to adjust this parameter, we can consider removing it from the signature in `marlin_gemm`.

Reply:
@zhyncs I have refactored the workspace code a bit based on @efrantar. Tell us if you see any other outstanding issue here.